In [ ]:
!pip install tensorflow-gpu==2.0.0-alpha0

In [ ]:
import numpy as np
import pandas as pd
import scipy
import os

from progressbar import ProgressBar

from sklearn.preprocessing import Normalizer

import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')

import tensorflow as tf
from tensorflow import keras

from sklearn.preprocessing import MinMaxScaler, Normalizer, StandardScaler

In [ ]:
tf.__version__

'2.0.0-alpha0'

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive', force_remount=False)

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


# DATA:

In [ ]:
X_train_path = 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/inertial_signals/'
X_test_path  = 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/inertial_signals/'
y_train_path = 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/y_train.txt'
y_test_path  = 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/y_test.txt'

In [ ]:
def get_all_file_paths(directory):

    # initializing empty file paths list
    file_paths = []

    # crawling through directory and subdirectories
    for root, directories, files in os.walk(directory):
        for filename in files:
            # join the two strings in order to form the full filepath.
            filepath = os.path.join(root, filename).replace("\\",'/')
            file_paths.append(filepath)

    # returning all file paths
    return file_paths

In [ ]:
all_train_paths = get_all_file_paths(X_train_path)
all_train_paths

['gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/inertial_signals/body_acc_x_train.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/inertial_signals/body_acc_y_train.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/inertial_signals/body_acc_z_train.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/inertial_signals/body_gyro_x_train.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/inertial_signals/body_gyro_y_train.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/inertial_signals/body_gyro_z_train.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/train/inertial_signals/total_acc_x_train.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_AC

In [ ]:
all_test_paths = get_all_file_paths(X_test_path)
all_test_paths

['gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/inertial_signals/body_acc_x_test.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/inertial_signals/body_acc_y_test.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/inertial_signals/body_acc_z_test.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/inertial_signals/body_gyro_x_test.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/inertial_signals/body_gyro_y_test.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/inertial_signals/body_gyro_z_test.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNIZE/data/UCI_HAR_Dataset/test/inertial_signals/total_acc_x_test.txt',
 'gdrive/My Drive/2_AAI/7_CASE_STUDIES_AAI/HUMAN_ACTIVITY_RECOGNI

In [ ]:
train_arry_1 = pd.read_csv(get_all_file_paths(X_train_path)[0], delim_whitespace=True, header=None).values
train_arry_2 = pd.read_csv(get_all_file_paths(X_train_path)[1], delim_whitespace=True, header=None).values
train_arry_3 = pd.read_csv(get_all_file_paths(X_train_path)[2], delim_whitespace=True, header=None).values
train_arry_4 = pd.read_csv(get_all_file_paths(X_train_path)[3], delim_whitespace=True, header=None).values
train_arry_5 = pd.read_csv(get_all_file_paths(X_train_path)[4], delim_whitespace=True, header=None).values
train_arry_6 = pd.read_csv(get_all_file_paths(X_train_path)[5], delim_whitespace=True, header=None).values
train_arry_7 = pd.read_csv(get_all_file_paths(X_train_path)[6], delim_whitespace=True, header=None).values
train_arry_8 = pd.read_csv(get_all_file_paths(X_train_path)[7], delim_whitespace=True, header=None).values
train_arry_9 = pd.read_csv(get_all_file_paths(X_train_path)[8], delim_whitespace=True, header=None).values

test_arry_1  = pd.read_csv(get_all_file_paths(X_test_path)[0], delim_whitespace=True, header=None).values
test_arry_2  = pd.read_csv(get_all_file_paths(X_test_path)[1], delim_whitespace=True, header=None).values
test_arry_3  = pd.read_csv(get_all_file_paths(X_test_path)[2], delim_whitespace=True, header=None).values
test_arry_4  = pd.read_csv(get_all_file_paths(X_test_path)[3], delim_whitespace=True, header=None).values
test_arry_5  = pd.read_csv(get_all_file_paths(X_test_path)[4], delim_whitespace=True, header=None).values
test_arry_6  = pd.read_csv(get_all_file_paths(X_test_path)[5], delim_whitespace=True, header=None).values
test_arry_7  = pd.read_csv(get_all_file_paths(X_test_path)[6], delim_whitespace=True, header=None).values
test_arry_8  = pd.read_csv(get_all_file_paths(X_test_path)[7], delim_whitespace=True, header=None).values
test_arry_9  = pd.read_csv(get_all_file_paths(X_test_path)[8], delim_whitespace=True, header=None).values

### FN THAT SHAPES DATA FOR LSTM USE:

In [ ]:
def retrn_lstm_data(listO_arrys):

    pbar = ProgressBar()

    def retrn_row_arry(row_idx, listO_arrys):

        listO_rows = []
        for arry in listO_arrys:
            row = arry[row_idx]
            listO_rows.append(row)
        row_arry = np.array(listO_rows).T

        return row_arry

    n_rows = listO_arrys[0].shape[0]
    listO_row_arrys = []
    for row_idx in pbar(range(n_rows)):

        row_arry = retrn_row_arry(row_idx, listO_arrys)
        listO_row_arrys.append(row_arry)

    arryO_row_arrys = np.array(listO_row_arrys)

    return arryO_row_arrys

In [ ]:
listO_train_arrys = [train_arry_1, train_arry_2, train_arry_3, train_arry_4, train_arry_5, train_arry_6, train_arry_7, train_arry_8, train_arry_9]
listO_test_arrys  = [test_arry_1, test_arry_2, test_arry_3, test_arry_4, test_arry_5, test_arry_6, test_arry_7, test_arry_8, test_arry_9]

### X_TRAIN:

In [ ]:
lstm_X_train = retrn_lstm_data(listO_train_arrys)
lstm_X_train.shape

100% (7352 of 7352) |####################| Elapsed Time: 0:00:00 Time:  0:00:00


(7352, 128, 9)

### X_TEST:

In [ ]:
lstm_X_test = retrn_lstm_data(listO_test_arrys)
lstm_X_test.shape

100% (2947 of 2947) |####################| Elapsed Time: 0:00:00 Time:  0:00:00


(2947, 128, 9)

### Y_TRAIN:

In [ ]:
lstm_y_train = pd.read_csv(y_train_path, delim_whitespace=True, header=None).values
lstm_y_train = pd.get_dummies(lstm_y_train.flatten()).values
lstm_y_train.shape

(7352, 6)

### Y_TEST:

In [ ]:
lstm_y_test = pd.read_csv(y_test_path, delim_whitespace=True, header=None).values
lstm_y_test = pd.get_dummies(lstm_y_test.flatten()).values
lstm_y_test.shape

(2947, 6)

In [ ]:
lstm_y_test

array([[0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1, 0],
       ...,
       [0, 1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0, 0]], dtype=uint8)

### DATA GENERATOR:

In [ ]:
def gen_batches(X, y, gen_constrain = False, n_EPOCHS = 2,
                      batch_size = 32):

    idxs = list(range(len(X)))

    c = 0
    while True:

        arryO_random_idxs = np.random.choice(idxs, batch_size, replace=False) #--------------------- SAMPLE WITHOUT REPLACEMENT

        X_batch = X[arryO_random_idxs]
        y_batch = y[arryO_random_idxs]

        yield X_batch, y_batch

        if gen_constrain != False:
            c += 1
            if c >= n_EPOCHS:
                break

### FN THAT TRAIN & TESTS LSTM MODELS:

In [ ]:
def fn_train_test_lstm(lstm_model, X_train, y_train, X_test, y_test, lr, batch_size, epochs, verbose = 1, model_save_path = None):

    optimizer = keras.optimizers.RMSprop(lr=lr, rho=0.9, epsilon=None, decay=0.0)
    lstm_model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

    # DATA GENERATOR SPECS:
    train_data_generator = gen_batches(lstm_X_train, lstm_y_train, batch_size = batch_size)
    val_data_generator   = gen_batches(lstm_X_test , lstm_y_test, batch_size = batch_size)

    n_train_data_pts = len(lstm_X_train)
    n_test_data_pts  = len(lstm_X_test)
    steps_per_epoch  = n_train_data_pts//batch_size
    validation_steps = n_test_data_pts//batch_size

    # MODEL CALLBACKS:
    if model_save_path != None:
        save_wts   = keras.callbacks.ModelCheckpoint(filepath = model_save_path, monitor = 'val_accuracy', save_best_only = True, mode = 'max')
    reduce_lr  = keras.callbacks.ReduceLROnPlateau(monitor = 'val_accuracy', factor = 0.1, patience = 5)
    callbacks_list = [save_wts, reduce_lr] if model_save_path != None else [reduce_lr]

    # FIT MODEL:
    model_history = lstm_model.fit_generator(train_data_generator, validation_data = val_data_generator, verbose = verbose,
                                             steps_per_epoch = steps_per_epoch, validation_steps = validation_steps,
                                             epochs=epochs, callbacks = callbacks_list)


    return model_history, lstm_model


### FN THAT GRID  SEARCHS ACROSS WEIGHT INITIALIZERS:

In [ ]:
def fn_grid_serach_lstms(fn_lstm_model, X_train, y_train, X_test, y_test, listO_kernel_initializers, lr, epochs, verbose = 0):

    pbar = ProgressBar()

    listO_performances = []
    c = 0
    BEST_val_acc = 0.80
    best_lstm = None
    for kernel_initializer in pbar(listO_kernel_initializers):

        lstm_model = fn_lstm_model(kernel_initializer)
        batch_size = 32
        model_history, lstm_model = fn_train_test_lstm(lstm_model, X_train, y_train, X_test, y_test, lr, batch_size, epochs, verbose = verbose)

        best_val_acc = max(model_history.history['val_accuracy'])
        epoch_n_best_acc = model_history.history['val_accuracy'].index(best_val_acc)
        corres_train_acc = model_history.history['accuracy'][epoch_n_best_acc]
        best_train_acc = max(model_history.history['accuracy'])
        performance = [kernel_initializer, best_val_acc, epoch_n_best_acc, corres_train_acc, best_train_acc]
        listO_performances.append(performance)

        if best_val_acc > BEST_val_acc:
            best_lstm_model = lstm_model
            BEST_val_acc = best_val_acc
        c += 1

    arryO_performances = np.array(listO_performances)
    columns = ['kernel_initializer', 'best_val_acc', 'epoch_n', 'corres_train_acc', 'best_train_acc']
    df_performance = pd.DataFrame(arryO_performances, columns = columns)

    return df_performance

# MODEL_1:

In [ ]:
def fn_lstm_model_1(kernel_initializer):

    n_time_steps = 128
    n_dims_per_timestep = 9
    n_classes = 6
    input_shape = (n_time_steps, n_dims_per_timestep)

    kwargs = dict(kernel_initializer=kernel_initializer)

    model_lstm = keras.models.Sequential()
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(64, input_shape=input_shape, return_sequences=True, **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(32, return_sequences=False,  **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))


    model_lstm.add(keras.layers.Dense(n_classes, activation = 'softmax'))

    return model_lstm

In [ ]:
fn_lstm_model =  fn_lstm_model_1
X_train, y_train =  lstm_X_train, lstm_y_train
X_test, y_test = lstm_X_test, lstm_y_test
listO_kernel_initializers = 'lecun_uniform glorot_normal he_normal lecun_normal he_uniform'.split()
lr, epochs = 1e-3, 200


df_performance = fn_grid_serach_lstms(fn_lstm_model, X_train, y_train, X_test, y_test, listO_kernel_initializers, lr, epochs)
df_performance

100% (5 of 5) |##########################| Elapsed Time: 1:19:52 Time:  1:19:52


kernel_initializer best_val_acc epoch_n corres_train_acc best_train_acc
0      lecun_uniform    0.9205163     137        0.9619269     0.96602076
1      glorot_normal    0.9351223     136        0.9507369     0.96001637
2          he_normal    0.9358016      23       0.96602076      0.9721616
3       lecun_normal    0.8974185      13       0.96165395     0.96834064
4         he_uniform     0.922894      15        0.9567413      0.9737991

#MODEL_2:

In [ ]:
def fn_lstm_model_2(kernel_initializer):

    n_time_steps = 128
    n_dims_per_timestep = 9
    n_classes = 6
    input_shape = (n_time_steps, n_dims_per_timestep)

    kwargs = dict(kernel_initializer=kernel_initializer)

    model_lstm = keras.models.Sequential()
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(12, input_shape=input_shape, return_sequences=True, **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(6, return_sequences=False,  **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))


    model_lstm.add(keras.layers.Dense(n_classes, activation = 'softmax'))

    return model_lstm

In [ ]:
fn_lstm_model =  fn_lstm_model_2
X_train, y_train =  lstm_X_train, lstm_y_train
X_test, y_test = lstm_X_test, lstm_y_test
listO_kernel_initializers = 'lecun_uniform glorot_normal he_normal lecun_normal he_uniform'.split()
lr, epochs = 1e-3, 150


df_performance = fn_grid_serach_lstms(fn_lstm_model, X_train, y_train, X_test, y_test, listO_kernel_initializers, lr, epochs)
df_performance

100% (5 of 5) |##########################| Elapsed Time: 1:00:02 Time:  1:00:02


kernel_initializer best_val_acc epoch_n corres_train_acc best_train_acc
0      lecun_uniform    0.9116848      75       0.91771287      0.9299945
1      glorot_normal    0.9123641      69        0.9366812      0.9407751
2          he_normal    0.8909647      11       0.87554586     0.92003274
3       lecun_normal     0.922894     140       0.92371726      0.9327238
4         he_uniform    0.9140625     112        0.9450055       0.950191

# MODEL_3:

In [ ]:
def fn_lstm_model_3(kernel_initializer):

    n_time_steps = 128
    n_dims_per_timestep = 9
    n_classes = 6
    input_shape = (n_time_steps, n_dims_per_timestep)

    kwargs = dict(kernel_initializer=kernel_initializer)

    model_lstm = keras.models.Sequential()
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(64, input_shape=input_shape, return_sequences=True, **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(32, return_sequences=True,  **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(32, return_sequences=False, **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))


    model_lstm.add(keras.layers.Dense(12, activation = 'relu'))
    model_lstm.add(keras.layers.Dense(n_classes, activation = 'softmax'))

    return model_lstm

In [ ]:
fn_lstm_model =  fn_lstm_model_3
X_train, y_train =  lstm_X_train, lstm_y_train
X_test, y_test = lstm_X_test, lstm_y_test
listO_kernel_initializers = 'lecun_uniform glorot_normal he_normal lecun_normal he_uniform'.split()
lr, epochs = 1e-3, 150


df_performance = fn_grid_serach_lstms(fn_lstm_model, X_train, y_train, X_test, y_test, listO_kernel_initializers, lr, epochs)
df_performance

100% (5 of 5) |##########################| Elapsed Time: 1:24:50 Time:  1:24:50


kernel_initializer best_val_acc epoch_n corres_train_acc best_train_acc
0      lecun_uniform    0.9303669      21        0.9567413     0.96247274
1      glorot_normal    0.9286685      53        0.9534662     0.95987993
2          he_normal   0.93444294      27        0.9586517      0.9619269
3       lecun_normal   0.93410325      67       0.95756006      0.9637009
4         he_uniform    0.9205163      39        0.9542849     0.96233624

# MODEL_4:

In [ ]:
def fn_lstm_model_4(kernel_initializer):

    n_time_steps = 128
    n_dims_per_timestep = 9
    n_classes = 6
    input_shape = (n_time_steps, n_dims_per_timestep)

    kwargs = dict(kernel_initializer=kernel_initializer)

    model_lstm = keras.models.Sequential()
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(32, input_shape=input_shape, return_sequences=True, **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(32,  return_sequences=True,  **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(50,  return_sequences=True,  **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(32,  return_sequences=True,  **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(keras.layers.Dropout(0.2))
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(32,  return_sequences=True,  **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())
    model_lstm.add(tf.compat.v1.keras.layers.CuDNNLSTM(12,  return_sequences=False, **kwargs))
    model_lstm.add(keras.layers.BatchNormalization())

    model_lstm.add(keras.layers.Dense(n_classes, activation = 'softmax'))

    return model_lstm

In [ ]:
fn_lstm_model =  fn_lstm_model_4
X_train, y_train =  lstm_X_train, lstm_y_train
X_test, y_test = lstm_X_test, lstm_y_test
listO_kernel_initializers = 'lecun_uniform glorot_normal he_normal lecun_normal he_uniform'.split()
lr, epochs = 1e-4, 150


df_performance = fn_grid_serach_lstms(fn_lstm_model, X_train, y_train, X_test, y_test, listO_kernel_initializers, lr, epochs)
df_performance

100% (5 of 5) |##########################| Elapsed Time: 2:27:28 Time:  2:27:28


kernel_initializer best_val_acc epoch_n corres_train_acc best_train_acc
0      lecun_uniform   0.92493206      38       0.94868994      0.9566048
1      glorot_normal   0.94225544     104        0.9459607       0.956059
2          he_normal    0.9001359     149       0.93299675     0.93545306
3       lecun_normal   0.92527175      51       0.94555134      0.9521015
4         he_uniform    0.8879076      86        0.9313592      0.9372271

# OBSERVATIONS:
- MODEL_4 USING GLOROT_NORMAL WT INITIALIZER GIVES THE BEST PERFORMANCE (94.22%) AMONG ALL THE LSTM MODELS IMPLEMETED
- MODELS 1 & 3 ALSO GIVE COMPARABLE PERFORMANCES, BUT GENERALIZATION NOT AS GOOD AS MODEL_4
- GENERALLY MOST OF THE LSTM MODELS CAN REACH A TESTSET PERFORMANCE OF AROUND 92%


### >> NEXT NOTEBOOK USE LSTM_MODEL_4 WITH GLOROT_NORMAL WT INITIALIZER & INSPECT CONFUSION MATRIX.